cwltool/provenance.py

```diff
@@ -0,0 +1,1184 @@
+
+            else:  # add the actual data value in the prov document
+                # Convert to bytes so we can get a hash (and add to RO)
+                b = io.BytesIO(str(value).encode(ENCODING))
+                data_file = self.add_data_file(b)
+                # FIXME: Don't naively assume add_data_file uses hash in filename!
+                data_id="data:%s" % posixpath.split(data_file)[1]
+                document.entity(data_id, {provM.PROV_TYPE:WFPROV["Artifact"], provM.PROV_VALUE:str(value)})
+                document.used(ProcessRunID, data_id, datetime.datetime.now(),None, {"prov:role":provRole })
+
+    def copy_job_order(self, r, job_order_object):
+        # type: (Any,Any) -> Any
+        '''
+        creates copy of job object for provenance
+        '''
+        if not hasattr(r, "tool"):
+            # direct command line tool execution
+            return job_order_object
+        customised_job={} #new job object for RO
+        for e, i in enumerate(r.tool["inputs"]):
+            with SourceLine(r.tool["inputs"], e, WorkflowException, _logger.isEnabledFor(logging.DEBUG)):
+                iid = shortname(i["id"])
+                if iid in job_order_object:
+                    customised_job[iid]= copy.deepcopy(job_order_object[iid]) #add the input element in dictionary for provenance
+                elif "default" in i:
+                    customised_job[iid]= copy.deepcopy(i["default"]) #add the defualt elements in the dictionary for provenance
+                else:
+                    raise WorkflowException(
+                        u"Input '%s' not in input object and does not have a default value." % (i["id"]))
+        return customised_job
+
+    def prospective_prov(self, document, r):
+        # type: (ProvDocument,Any) -> None
+        '''
+        create prospective provenance recording for the workflow as wfdesc prov:Plan
+        '''
+        if not hasattr(r, "steps"):
+            # direct command line tool execution
+            document.entity("wf:main", {provM.PROV_TYPE: WFDESC["Process"], "prov:type": PROV["Plan"], "prov:label":"Prospective provenance"})
+            return
+
+        # FIXME: Workflow is always called "#main"?
+        document.entity("wf:main", {provM.PROV_TYPE: WFDESC["Workflow"], "prov:type": PROV["Plan"], "prov:label":"Prospective provenance"})
+
+        steps=[]
+        for s in r.steps:
+            # FIXME: Use URI fragment identifier for step name, e.g. for spaces
+            stepnametemp="wf:main/"+str(s.name)[5:]
+            stepname=urllib.parse.quote(stepnametemp, safe=":/,#")
+            steps.append(stepname)
+            step = document.entity(stepname, {provM.PROV_TYPE: WFDESC["Process"], "prov:type": PROV["Plan"]})
+            document.entity("wf:main", {"wfdesc:hasSubProcess":step, "prov:label":"Prospective provenance"})
+
+        # TODO: Declare roles/parameters as well
+
+
+#**************************************
```

Comment:
The prospective provenance does not currently declare the input/outputs of the steps/workflow - but we refer to those in the provenance with `prov:role`. See [wfdesc:Parameter](http://wf4ever.github.io/ro/2016-01-28/wfdesc/#Parameter)

In PROV datamodel the [role](https://www.w3.org/TR/2013/REC-prov-dm-20130430/#term-attribute-role) is not typed, but it is a [prov:Role](https://www.w3.org/TR/2013/REC-prov-o-20130430/#hadRole) in PROV-O.  In either case, we should declare it as part of the prospective provenance so that the link to the step is explicit and not just hidden in the URI.